modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EChain.java

```diff
@@ -233,6 +248,11 @@ private void analyzeWrite(final CompilerSettings settings, final Definition defi
     private void analyzeRead() {
         final ALink last = links.get(links.size() - 1);
 
+        // If the load node is a DEF node, we adapt its after type to use _this_ expected output type:
+        if (last instanceof ADefLink && this.expected != null) {
+            last.after = this.expected;
+        }
+
```

Comment:
No, this one should work - and works. My code is not involved here. The new code only involves def's pointing to arrays or calls on def's fields or methods. DefLink is the superclass of such nodes. Plain "def" variables are not affected, they involve no invokedynamic.